src/main/java/ch/njol/skript/expressions/ExprXYZComponent.java

```diff
@@ -0,0 +1,198 @@
+	"add 1 to x of {_v}",
+	"add 2 to y of {_v}",
+	"add 3 to z of {_v}",
+	"send \"%x of {_v}%, %y of {_v}%, %z of {_v}%\"",
+	"set x component of {_v} to 1",
+	"set y component of {_v} to 2",
+	"set z component of {_v} to 3",
+	"send \"%x component of {_v}%, %y component of {_v}%, %z component of {_v}%\""
+})
+@Since("2.2-dev28, INSERT VERSION (Quaternions)")
+public class ExprXYZComponent extends SimplePropertyExpression<Object, Number> {
+
+	static {
+		String types = "vectors";
+		if (Skript.isRunningMinecraft(1, 19, 4))
+			types += "/quaternions";
+		register(ExprXYZComponent.class, Number.class, "[vector|quaternion] (0¦w|1¦x|2¦y|3¦z) [component[s]]", types);
+	}
+
+	private final static Character[] axes = new Character[] {'w', 'x', 'y', 'z'};
+
+	private int axis;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		axis = parseResult.mark;
+		return super.init(exprs, matchedPattern, isDelayed, parseResult);
+	}
+
+	@Override
+	public Number convert(Object object) {
+		if (object instanceof Vector) {
+			Vector vector = (Vector) object;
+			return axis == 1 ? vector.getX() : (axis == 2 ? vector.getY() : vector.getZ());
+		} else {
+			Quaternionf quaternion = (Quaternionf) object;
+			switch (axis) {
+				case 0:
+					return quaternion.w();
+				case 1:
+					return quaternion.x();
+				case 2:
+					return quaternion.y();
+				case 3:
+					return quaternion.z();
+				default:
+					return null;
+			}
+		}
+	}
+
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (getExpr().getReturnType().equals(Quaternionf.class)) {
+			if ((mode == ChangeMode.SET || mode == ChangeMode.ADD || mode == ChangeMode.REMOVE))
+				return new Class[] {Number.class};
+		}
+		if ((mode == ChangeMode.ADD || mode == ChangeMode.REMOVE || mode == ChangeMode.SET)
+				&& getExpr().isSingle() && Changer.ChangerUtils.acceptsChange(getExpr(), ChangeMode.SET, Vector.class))
+			return CollectionUtils.array(Number.class);
+		return null;
+	}
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		assert delta != null;
+		for (Object object : getExpr().getArray(event)) {
+			if (object instanceof Vector) {
+				Vector vector = (Vector) object;
+				double value = ((Number) delta[0]).doubleValue();
+				switch (mode) {
+					case REMOVE:
+						value = -value;
+						//$FALL-THROUGH$
+					case ADD:
+						if (axis == 0)
+							vector.setX(vector.getX() + value);
+						else if (axis == 1)
+							vector.setY(vector.getY() + value);
+						else
+							vector.setZ(vector.getZ() + value);
+						getExpr().change(event, new Vector[] {vector}, ChangeMode.SET);
+						break;
+					case SET:
+						if (axis == 0)
+							vector.setX(value);
+						else if (axis == 1)
+							vector.setY(value);
+						else
+							vector.setZ(value);
+						getExpr().change(event, new Vector[] {vector}, ChangeMode.SET);
```

Comment:
```suggestion
						if (axis == 1) {
							vector.setX(vector.getX() + value);
						} else if (axis == 2) (
							vector.setY(vector.getY() + value);
						} else {
							vector.setZ(vector.getZ() + value);
						}
						getExpr().change(event, new Vector[] {vector}, ChangeMode.SET);
						break;
					case SET:
						if (axis == 1) {
							vector.setX(value);
						} else if (axis == 2) {
							vector.setY(value);
						} else {
							vector.setZ(value);
						}
						getExpr().change(event, new Vector[] {vector}, ChangeMode.SET);
```
This plus something to catch axis = 0 earlier on.

Reply:
Still needs brackets

Reply:
Old implementation